File: data/src/app/member.kt
package data.app

import androidx.room.ColumnInfo
import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.PrimaryKey
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow
import kotlinx.serialization.Serializable

@Entity(
    tableName = "workspace_members",
    primaryKeys = ["workspace_id", "user_id"],
    foreignKeys = [
        ForeignKey(
            entity = Workspace::class,
            parentColumns = ["id"],
            childColumns = ["workspace_id"],
            onDelete = ForeignKey.CASCADE
        ),
        ForeignKey(
            entity = User::class,
            parentColumns = ["id"],
            childColumns = ["user_id"],
            onDelete = ForeignKey.CASCADE
        )
    ],
    indices = [Index("workspace_id"), Index("user_id")]
) data class Member(
    @ColumnInfo(name = "workspace_id") val workspaceId: String,
    @ColumnInfo(name = "user_id") val userId: String,
    val role: MemberRole = MemberRole.MEMBER,
    @ColumnInfo(name = "joined_at", defaultValue = "CURRENT_TIMESTAMP") val joinedAt: String = ""
)

@Serializable enum class MemberRole { OWNER, ADMIN, MEMBER, GUEST }

@Dao interface MemberDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE) suspend fun upsert(item: Member)
    @Update suspend fun update(item: Member)
    @Delete suspend fun delete(item: Member)

    @Query("SELECT * FROM workspace_members WHERE workspace_id = :workspaceId")
    fun getByWorkspaceId(workspaceId: String): Flow<List<Member>>

    @Query("SELECT * FROM workspace_members WHERE user_id = :userId")
    fun getByUserId(userId: String): Flow<List<Member>>

    @Query("SELECT * FROM workspace_members WHERE workspace_id = :workspaceId AND user_id = :userId")
    suspend fun get(workspaceId: String, userId: String): Member?

    @Query("SELECT * FROM workspace_members")
    fun getAll(): Flow<List<Member>>
}

File: server/data/app/member.ts
import { sqliteTable, text, primaryKey, index } from "drizzle-orm/sqlite-core"
import { eq, sql, and } from "drizzle-orm"
import { DrizzleD1Database } from "drizzle-orm/d1"
import { User } from "./user"
import { Workspace } from "./workspace"

export const MemberRoleValues = ["OWNER", "ADMIN", "MEMBER", "GUEST"] as const
export type MemberRole = (typeof MemberRoleValues)[number]

export const Member = sqliteTable(
  "workspace_members",
  {
    joined_at: text("joined_at").default(sql`CURRENT_TIMESTAMP`),
    role: text("role"),
    user_id: text("user_id").references(() => User.id, { onDelete: "cascade" }),
    workspace_id: text("workspace_id").references(() => Workspace.id, { onDelete: "cascade" }),
  },
  (table) => ({
    pk: primaryKey({ columns: [table.workspace_id, table.user_id] }),
    idx0: index("workspace_members_workspace_id_idx").on(table.workspace_id),
    idx1: index("workspace_members_user_id_idx").on(table.user_id),
  }),
)

export class MemberDao {
  constructor(private db: DrizzleD1Database<any>) {}

  async upsert(item: typeof Member.$inferInsert) {
    return this.db
      .insert(Member)
      .values(item)
      .onConflictDoUpdate({ target: [Member.workspace_id, Member.user_id], set: item })
      .run()
  }

  async update(item: Partial<typeof Member.$inferInsert> & { workspace_id: any; user_id: any }) {
    return this.db
      .update(Member)
      .set(item)
      .where(and(eq(Member.workspace_id, item.workspace_id!), eq(Member.user_id, item.user_id!)))
      .run()
  }

  async delete(item: { workspace_id: any; user_id: any }) {
    return this.db
      .delete(Member)
      .where(and(eq(Member.workspace_id, item.workspace_id!), eq(Member.user_id, item.user_id!)))
      .run()
  }

  async get(workspace_id: any, user_id: any) {
    return this.db
      .select()
      .from(Member)
      .where(and(eq(Member.workspace_id, workspace_id), eq(Member.user_id, user_id)))
      .get()
  }

  async getAll() {
    return this.db.select().from(Member).all()
  }
}
